openmc/dagmc.py

```diff
@@ -0,0 +1,620 @@
+        self._auto_mat_ids = val
+
+    @property
+    def material_names(self):
+        dagmc_file_contents = h5py.File(self.filename)
+        material_tags_hex = dagmc_file_contents['/tstt/tags/NAME'].get(
+            'values')
+        material_tags_ascii = []
+        for tag in material_tags_hex:
+            candidate_tag = tag.tobytes().decode().replace('\x00', '')
+            # tags might be for temperature or reflective surfaces
+            if candidate_tag.startswith('mat:'):
+                # if name ends with _comp remove it, it is not parsed
+                if candidate_tag.endswith('_comp'):
+                   candidate_tag = candidate_tag[:-5]
+                # removes first 4 characters as openmc.Material name should be
+                # set without the 'mat:' part of the tag
+                material_tags_ascii.append(candidate_tag[4:])
+
+        return sorted(set(material_tags_ascii))
+
+    def _n_geom_elements(self, geom_type):
+        """
+        Helper function for retrieving the number geometric entities in a DAGMC
+        file
+
+        Parameters
+        ----------
+        geom_type : str
+            The type of geometric entity to count. One of {'Volume', 'Surface'}. Returns
+            the runtime number of voumes in the DAGMC model (includes implicit complement).
+
+        Returns
+        -------
+        int
+            Number of geometry elements of the specified type
+        """
+        cv.check_value('geometry type', geom_type, ('volume', 'surface'))
+
+        def decode_str_tag(tag_val):
+            return tag_val.tobytes().decode().replace('\x00', '')
+
+        with h5py.File(self.filename) as dagmc_file:
+            category_data = dagmc_file['tstt/tags/CATEGORY/values']
+            category_strs = map(decode_str_tag, category_data)
+            n = sum([v == geom_type.capitalize() for v in category_strs])
+
+            # check for presence of an implicit complement in the file and
+            # increment the number of cells if it doesn't exist
+            if geom_type == 'volume':
+                name_data = dagmc_file['tstt/tags/NAME/values']
+                name_strs = map(decode_str_tag, name_data)
+                if not sum(['impl_complement' in n for n in name_strs]):
+                    n += 1
+        return n
+
+    @property
+    def n_cells(self):
+        return self._n_geom_elements('volume')
+
+    @property
+    def n_surfaces(self):
+        return self._n_geom_elements('surface')
+
+    def create_xml_subelement(self, xml_element, memo=None):
+        if memo is None:
+            memo = set()
+
+        if self in memo:
+            return
+
+        memo.add(self)
+
+        # Ensure that the material overrides are up-to-date
+        for cell in self.cells.values():
+            if cell.fill is None:
+                continue
+            fill = cell.fill if isinstance(cell.fill, openmc.Iterable) else [cell.fill]
```

Comment:
```suggestion
            fill = cell.fill
```
This should be enough as `add_material_override()` ensures an Iterable to be added to the dictionary.

Reply:
cc @pshriwise

Reply:
Agreed 👍🏻